src/main/java/io/pravega/perf/PerfStats.java

```diff
@@ -5,227 +5,363 @@
-    final private long windowInterval;
-    private timeWindow window;
-    final private AtomicLong eventID;
-    final private CSVPrinter printer;
-
+    final private String csvFile;
+    final private int messageSize;
+    final private int windowInterval;
+    final private ConcurrentLinkedQueue<TimeStamp> queue;
+    final private ForkJoinPool executor;
+    private QueueProcessor recorder;
 
     /**
-     * private class for Performance statistics within a given time window.
+     * private class for start and end time.
      */
-    private class timeWindow {
-        final private Instant startTime;
-        private Instant lastTime;
-        private long count;
-        private long bytes;
-        private long maxLatency;
-        private long totalLatency;
-
-        public timeWindow() {
-            this.startTime = Instant.now();
-            this.lastTime = this.startTime;
-            this.count = 0;
-            this.bytes = 0;
-            this.maxLatency = 0;
-            this.totalLatency = 0;
+    final static private class TimeStamp {
+        final private long startTime;
+        final private long endTime;
+        final private int bytes;
+
+        private TimeStamp(long startTime, long endTime, int bytes) {
+            this.startTime = startTime;
+            this.endTime = endTime;
+            this.bytes = bytes;
         }
+    }
 
-        /**
-         * record the latency and bytes
-         *
-         * @param latency latency in ms.
-         * @param bytes   number of bytes.
-         */
-        public void record(long latency, long bytes) {
-            this.count++;
-            this.totalLatency += latency;
-            this.bytes += bytes;
-            this.maxLatency = Math.max(this.maxLatency, latency);
+
+    public PerfStats(String action, int reportingInterval, int messageSize, String csvFile) {
+        this.action = action;
+        this.messageSize = messageSize;
+        this.windowInterval = reportingInterval;
+        this.csvFile = csvFile;
+        this.queue = new ConcurrentLinkedQueue<TimeStamp>();
+        this.executor = new ForkJoinPool(1);
+        this.recorder = null;
+    }
+
+    /**
+     * private class for start and end time.
+     */
+    @NotThreadSafe
+    final static private class QueueProcessor implements Runnable {
+        final private String action;
+        final private LatencyWriter latencyRecorder;
+        final private long windowInterval;
+        final private ConcurrentLinkedQueue<TimeStamp> queue;
+
+
+        private QueueProcessor(String action, int reportingInterval, int messageSize,
+                               long startTime, String csvFile, ConcurrentLinkedQueue<TimeStamp> queue) throws IOException {
+            this.action = action;
+            this.windowInterval = reportingInterval;
+            this.queue = queue;
+            this.latencyRecorder = csvFile == null ? new LatencyWriter(action, messageSize, startTime) :
```

Comment:
If this instantiation were delayed until run(), and instead 'action, messageSize, and startTime' were final member variables, then this class would be thread safe.

Reply:
ok, its done

Reply:
I do not see this change.

Reply:
now , I have made the Latency writer as a local variable to queue processor thread.
to stop this thread, I will be sending end msg to queue.   I remember that last time you commented  that a sending end msg looks odd; but , I need full latency calculations at the end of the bench-marking and i dont want to use locking here . so, this approach should be ok with respect to achieving thread safety by not using locks.